pkg/krok/providers/auth/oauth.go

```diff
@@ -0,0 +1,149 @@
+	UUID   providers.UUIDGenerator
+	Clock  providers.Clock
+	Issuer providers.UserTokenIssuer
+}
+
+// OAuthAuthenticator is the OAuthAuthenticator that uses OAuth2 to authenticate the user.
+type OAuthAuthenticator struct {
+	OAuthAuthenticatorConfig
+	OAuthAuthenticatorDependencies
+
+	// TODO: Have a map of configs and support multiple providers.
+	oauthCfg *oauth2.Config
+}
+
+// NewOAuthAuthenticator creates a new OAuthAuthenticator.
+func NewOAuthAuthenticator(cfg OAuthAuthenticatorConfig, deps OAuthAuthenticatorDependencies) *OAuthAuthenticator {
+	return &OAuthAuthenticator{
+		OAuthAuthenticatorConfig:       cfg,
+		OAuthAuthenticatorDependencies: deps,
+
+		// For now, just support Google.
+		oauthCfg: &oauth2.Config{
+			ClientID:     cfg.GoogleClientID,
+			ClientSecret: cfg.GoogleClientSecret,
+			RedirectURL:  "http://localhost:9998/auth/callback",
+			Scopes: []string{
+				"https://www.googleapis.com/auth/userinfo.email",
+				"https://www.googleapis.com/auth/userinfo.profile",
+			},
+			Endpoint: google.Endpoint,
+		},
+	}
+}
+
+// GetAuthCodeURL gets the OAuth2 authentication URL.
+func (op *OAuthAuthenticator) GetAuthCodeURL(state string) string {
+	return op.oauthCfg.AuthCodeURL(state, []oauth2.AuthCodeOption{oauth2.AccessTypeOffline}...)
+}
+
+// Exchange exchanges the code returned from the OAuth2 authentication URL for a valid token.
+// We then call the TokenIssuer to get an internal JWT.
+func (op *OAuthAuthenticator) Exchange(ctx context.Context, code string) (*oauth2.Token, error) {
+	token, err := op.oauthCfg.Exchange(ctx, code)
+	if err != nil {
+		return nil, err
+	}
+
+	userAuthDetails, err := op.getGoogleUser(token.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+
+	return op.Issuer.Create(ctx, userAuthDetails)
+}
+
+// stateClaims are used when creating a temporary JWT state nonce that has an expiry.
+// This is used for CSRF protection when logging in via an OAuth2 provider.
+type stateClaims struct {
+	jwt.StandardClaims
+	RedirectURL string `json:"redirect_url"`
+}
+
+// GenerateState generates the state nonce JWT with expiry.
+func (op *OAuthAuthenticator) GenerateState(redirectURL string) (string, error) {
+	uuid, err := op.UUID.Generate()
+	if err != nil {
+		return "", fmt.Errorf("uuid generate: %w", err)
+	}
+
+	now := op.Clock.Now()
+	claims := stateClaims{
+		StandardClaims: jwt.StandardClaims{
+			Subject:   uuid,
+			ExpiresAt: now.Add(time.Minute * 2).Unix(),
+			IssuedAt:  now.Unix(),
+		},
+		RedirectURL: redirectURL,
+	}
+	state, err := jwt.NewWithClaims(jwt.SigningMethodHS256, claims).SignedString([]byte(op.GlobalTokenKey))
+	if err != nil {
+		return "", fmt.Errorf("new signed token: %w", err)
+	}
+
+	return state, nil
+}
+
+// VerifyState verifies the state nonce JWT.
+func (op *OAuthAuthenticator) VerifyState(rawToken string) (string, error) {
+	var claims stateClaims
+	_, err := jwt.ParseWithClaims(rawToken, &claims, func(token *jwt.Token) (interface{}, error) {
```

Comment:
if _, err := ... short form.

Also, don't forget to add logging on error cases. :)